Include dungeon name in hinted keys

mracsys opened the pull request:
Keeping in draft for now as previous discussion on Discord was unclear if this would be accepted.

Instead of location hints saying "X holds a Small Key", include the dungeon name: "X holds a Forest Temple Small Key". There is precedence with the Gerudo Fortress keys hinted as a jail key instead of a small key.

Applies to both small keys and boss keys, including the Ganon boss key in keysanity.

mracsys committed:
include dungeon name in hinted keys

mracsys committed:
fix a dumb typo

Hamsda commented:
No "the" for `"Ganons Castle Boss Key"` intentional?

mracsys commented:
Yep, it was intentional. It sounded better to me even if including "the" is technically right.

r0bd0g commented:
This implementation causes a significant difference in gameplay between normal and clearer hints that hasn't existed before. I think you need to come up with unique non-clearer hints for all key types.

fenhl commented on HintList.py:
```diff
@@ -240,6 +240,20 @@ def tokens_required_by_settings(world):
+    'Boss Key (Water Temple)':                                  (["a master of unlocking", "a dungeon's master pass"], "the Water Temple Boss Key", 'item'),
+    'Boss Key (Shadow Temple)':                                 (["a master of unlocking", "a dungeon's master pass"], "the Shadow Temple Boss Key", 'item'),
+    'Boss Key (Spirit Temple)':                                 (["a master of unlocking", "a dungeon's master pass"], "the Spirit Temple Boss Key", 'item'),
+    'Boss Key (Ganons Castle)':                                 (["a master of unlocking", "a dungeon's master pass"], "Ganons Castle Boss Key", 'item'),
```
```suggestion
    'Boss Key (Ganons Castle)':                                 (["a master of unlocking", "a dungeon's master pass"], "Ganon's Castle Boss Key", 'item'),
```

fenhl commented on HintList.py:
```diff
@@ -240,6 +240,20 @@ def tokens_required_by_settings(world):
+    'Small Key (Spirit Temple)':                                (["a tool for unlocking", "a dungeon pass", "a lock remover", "a lockpick"], "a Spirit Temple Small Key", 'item'),
+    'Small Key (Bottom of the Well)':                           (["a tool for unlocking", "a dungeon pass", "a lock remover", "a lockpick"], "a Bottom of the Well Small Key", 'item'),
+    'Small Key (Gerudo Training Ground)':                       (["a tool for unlocking", "a dungeon pass", "a lock remover", "a lockpick"], "a Gerudo Training Ground Small Key", 'item'),
+    'Small Key (Ganons Castle)':                                (["a tool for unlocking", "a dungeon pass", "a lock remover", "a lockpick"], "a Ganons Castle Small Key", 'item'),
```
```suggestion
    'Small Key (Ganons Castle)':                                (["a tool for unlocking", "a dungeon pass", "a lock remover", "a lockpick"], "a Ganon's Castle Small Key", 'item'),
```

mracsys committed:
adjust vague hint text for keys

mracsys committed:
add apostrophe to Ganons

mracsys commented:
I've addressed both r0b's and fenhl's comments.

r0bd0g commented:
I think these non-clearer hints might be a little much. I suggest keeping it simple. Maybe something "a master of unlocking for a ruined citadel", "a tool for unlocking on a high mountain", for examples?

(I kind of like not knowing what the keys are for. I suspect I'm in the minority here... I'll miss it a little though.)

mracsys commented:
You're right, keys should not have full names in their own dungeons. Maps/Compasses should also be in if shuffled outside their dungeons.

For vague hints, do you have suggestions for ER-independent descriptors? I could make them consistent with the easier choice instead of being unique (hot tool/pass/remover instead of torrid/hot/sizzling).

r0bd0g commented:
I suggested previously, combining the current small key non-clearer hints with the dungeon names (vast lake, high mountain, etc). I know that those aren't super ER-independent but it is what's currently in use for those dungeons regardless. (So example, "a master of unlocking in a ruined citadel".) So this combines things you already know for non-clearer, saving having to learn something new for each type of item. Having so many hints I think makes the non-clearer hints more confusing than they ought to be, remembering each hint for all the types of keys it a bit much, so I'd also recommend paring it back to just one hint text per key.

Something a bit strange is how a hint text will reveal the type of key, but you can't tell just by looking. There's precedent for this but it's a bit uncommon, with ice trap hints telling you they're ice traps. (I don't think coloured keys are a good idea though b/c I think that'll just make it more confusing to recognize what distant freestandings are.)

mracsys committed:
add maps/compasses, change vague names

mracsys commented:
I updated the vague hint text and added support for compasses and maps. Dungeon names are not included when the item is shuffled in its own dungeon or vanilla. Below is a screenshot with the longest location and item strings to verify it fits (barely).

![image](https://user-images.githubusercontent.com/19615534/131237939-fc430fd0-ea6b-4657-b989-9b1421aa931b.png)

r0bd0g commented:
Yeah, whether or not the longest texts fit was gonna be my next question. Tight squeeze there. I'm guessing MW doesn't say who the item is for? If it did then this text would definitely go too long. I wonder if the fact that it doesn't give with world ought to influence this PR. Imagine you see a bow hinted somewhere crappy, but it's the only available bow for player 75 or whatever. Kind of similar to seeing a key and not knowing if it's one you really need.

mracsys committed:
consistent forest temple vague hints

mracsys commented:
> I noticed in some cases you've worded it "for a deep forest" and others "for deep in a forest"?

That's what happens when I work on this stuff on little sleep 😄. It's fixed now to always be "for a deep forest".

For how multiworld compares with this change, the scope is larger. This change is making key specificity consistent with other hinted items. World info would be new info for every item. That feels like another conversation/PR.

Zannick commented on Item.py:
```diff
@@ -115,6 +115,12 @@ def compass(self):
+        return ((self.type == 'SmallKey' and self.world.settings.shuffle_smallkeys in ['remove','vanilla','dungeon']) or
+                (self.type == 'FortressSmallKey' and self.world.settings.shuffle_fortresskeys in ['vanilla']) or
+                (self.bosskey and self.world.settings.shuffle_bosskeys in ['remove','vanilla','dungeon']) or
+                (self.map or self.compass and self.world.settings.shuffle_mapcompass in ['remove','startwith','vanilla','dungeon']))
```
This line needs additional parens to clarify the effect. eg. `((map or compass) and (shuffle_mapcompass...))`

mracsys replied:
Implemented as requested

Zannick commented on HintList.py:
```diff
@@ -237,11 +237,46 @@ def tokens_required_by_settings(world):
+    'Small Key (Water Temple)':                                 (["a tool for unlocking a vast lake", "a dungeon pass for under a vast lake", "a lock remover for under a vast lake", "a lockpick for under a vast lake"], "a Water Temple Small Key", 'item'),
+    'Small Key (Shadow Temple)':                                (["a tool for unlocking the house of the dead", "a dungeon pass for the house of the dead", "a lock remover for the house of the dead", "a lockpick for the house of the dead"], "a Shadow Temple Small Key", 'item'),
+    'Small Key (Spirit Temple)':                                (["a tool for unlocking a goddess of the sand", "a dungeon pass for a goddess of the sand", "a lock remover for a goddess of the sand", "a lockpick for a goddess of the sand"], "a Spirit Temple Small Key", 'item'),
+    'Small Key (Bottom of the Well)':                           (["a tool for unlocking a shadow\'s prison", "a dungeon pass for a shadow\'s prison", "a lock remover for a shadow\'s prison", "a lockpick for a shadow\'s prison"], "a Bottom of the Well Small Key", 'item'),
```
The backslashes in `\'` are unnecessary when the string is wrapped in double-quotes.

mracsys replied:
`\` taken out. The hint text was copied from line 1229 if you want me to change that as well.

Zannick replied:
Go for it.

mracsys committed:
address code review comments

mracsys committed:
remove unnecessary backslash in hint text
